### ISSUE TITLE
Compiling against Berkeley DB SQL API

### ISSUE THREAD

**Issue by brianjpetersen** [rogerbinns/apsw#216]
I've been happily using apsw for interacting with sqlite for a few years now - thanks for developing an amazing piece of software!

I've run into a use-case where I may be pushing the boundaries of what sqlite is able to handle in terms of concurrency (several dozen concurrent writers mingled with periodic slow-running selects).  I've tried the recommended ways of squeezing more concurrency out of sqlite, and before I fall back on a more traditional solution for this problem such as postgres, I was hoping to try the approach described here with apsw:

http://charlesleifer.com/blog/updated-instructions-for-compiling-berkeleydb-with-sqlite-for-use-with-python/

Basically, the newer releases of Berkeley DB have an API which is claimed compatible with sqlite with vastly improved concurrency due to page-level locking as opposed to file-level locking.

Two questions:

1. Are there any reasons why attempting to compile apsw against the Berkeley DB API is a bad idea?
2. How can I accomplish this?  Is it simply a matter of including the appropriate header and compiled library in the directory sqlite3/ before running setup.py?

Thanks in advance for your help!

**Comment by rogerbinns** [comment]
You can try, but it is unlikely to be successful.  You need to ensure that something providing the SQLite API [can be found during building](https://rogerbinns.github.io/apsw/build.html#finding-sqlite-3).

However APSW wraps virtually the entire SQLite API, so you will have problems if the berkleydb version is not complete.  (Historically it hasn't been.)  You'll need to adjust the code on either side to your satisfaction.  Alternately you can compile pysqlite instead (which is what your link does) as pysqlite wraps far less of the SQLite API and supports very old versions of it.

However I'd expect that you can't do substantially better than SQLite's WAL mode.
